pandas/core/aggregation.py

```diff
@@ -415,67 +423,98 @@ def transform(
-            colg = obj._gotitem(name, ndim=1)
-            try:
-                results[name] = transform(colg, how, 0, *args, **kwargs)
-            except Exception as e:
-                if str(e) == "Function did not transform":
-                    raise e
-
-        # combine results
-        if len(results) == 0:
+    if is_dict_like(func):
+        func = cast(Dict[Label, Union[AggFuncTypeBase, List[AggFuncTypeBase]]], func)
+        result = transform_dict_like(obj, func, *args, **kwargs)
+    else:
+        func = cast(AggFuncTypeBase, func)
+        try:
+            result = transform_str_or_callable(obj, func, *args, **kwargs)
+        except Exception:
             raise ValueError("Transform function failed")
-        return concat(results, axis=1)
-
-    # func is either str or callable
-    try:
-        if isinstance(func, str):
-            result = obj._try_aggregate_string_function(func, *args, **kwargs)
-        else:
-            f = obj._get_cython_func(func)
-            if f and not args and not kwargs:
-                result = getattr(obj, f)()
-            else:
-                try:
-                    result = obj.apply(func, args=args, **kwargs)
-                except Exception:
-                    result = func(obj, *args, **kwargs)
-    except Exception:
-        raise ValueError("Transform function failed")
 
     # Functions that transform may return empty Series/DataFrame
     # when the dtype is not appropriate
     if isinstance(result, (ABCSeries, ABCDataFrame)) and result.empty:
         raise ValueError("Transform function failed")
+
     if not isinstance(result, (ABCSeries, ABCDataFrame)) or not result.index.equals(
         obj.index
     ):
         raise ValueError("Function did not transform")
 
     return result
+
+
+def transform_dict_like(
+    obj: FrameOrSeries,
+    func: Dict[Label, Union[AggFuncTypeBase, List[AggFuncTypeBase]]],
+    *args,
+    **kwargs,
+):
+    """
+    Compute transform in the case of a dict-like func
+    """
+    from pandas.core.reshape.concat import concat
+
+    if obj.ndim != 1:
+        # Check for missing columns on a frame
+        cols = sorted(set(func.keys()) - set(obj.columns))
+        if len(cols) > 0:
+            raise SpecificationError(f"Column(s) {cols} do not exist")
+
+    # Can't use func.values(); wouldn't work for a Series
+    if any(is_dict_like(v) for _, v in func.items()):
+        # GH 15931 - deprecation of renaming keys
+        raise SpecificationError("nested renamer is not supported")
+
+    if len(func) == 0:
+        raise ValueError("no results")
+
+    results: Dict[Label, FrameOrSeriesUnion] = {}
+    for name, how in func.items():
+        colg = obj._gotitem(name, ndim=1)
+        try:
+            results[name] = transform(colg, how, 0, *args, **kwargs)
+        except Exception as err:
+            if str(err) == "Function did not transform" or str(err) == "no results":
+                raise err
```

Comment:
Would it be reasonable to create a named exception rather than testing the error message match?